### ISSUE TITLE
Work on nodes instead of files?

### ISSUE THREAD

**Issue by alphapapa** [Kungsgeten/org-brain#10]
Hi,

As discussed at https://www.reddit.com/r/emacs/comments/630j3q/orgbrain_now_on_melpa/dfreedh/

Rather than using many small Org files, I use a few large ones, so if I were to use org-brain, I would need it to work by linked nodes/headings instead of linked files.

Is this possible or feasible?  I realize it might require a major refactoring.  :)

**Comment by Kungsgeten** [comment]
Hi! I think its a good idea, but I'm not quite sure how it would work. My guess is that each headline with a `CUSTOM_ID` should be treated as an `org-brain` entry, and that all entries (files or other ID headlines) linking to that headline should be considered its parents? The file containing the headline itself should probably be a parent too... Should subheadings be considered (children) entries, or just normal headlines?

The above concerns the "user interface", not how to program it. That could probably be done, though I'm not quite sure how. If we figure out a way how to write it in the `org`-files first, perhaps? :)

**Comment by alphapapa** [comment]
Yeah that's kind of a tough call, whether subheadings should be considered children.  In a way that would make sense, but on the other hand the whole point is to link headings together arbitrarily, outside of the rigid, single-parent outline hierarchy.  But back on the first hand, doing both seems like the best of both worlds, haha.

The one suggestion I would have is to, well, just ignore the `CUSTOM_ID`, because that requires users to go through an extra step.  The plain `id` property would work well enough, and is easily assigned with a built-in function.  But even then, is it necessary?  I would suggest to just consider any heading that is linked to by other headings or links to other headings to be an `org-brain` entry.  But then again, maybe you've already thought of that and decided better against it.  :)

**Comment by Kungsgeten** [comment]
Well, the main reason for using some kind of id links would probably be to make sure that the link links to a headline which actually exist. I believe the other way is to do like `[[*My headline]]` if I'm not mistaken? Then we'd have to parse those links, go to see if the headline actually exists etc.

**Comment by alphapapa** [comment]
Oh, I see what you mean.  Well, as far as I understand, there's no reason to use `CUSTOM_ID` over the `ID` field.  For example, that's what I use in `org-bookmark-heading`.

Well, here's what the info page says:

```
If the headline has a `CUSTOM_ID' property, a link to this custom
     ID will be stored.  In addition or alternatively (depending on the
     value of `org-id-link-to-org-use-id'), a globally unique `ID'
     property will be created and/or used to construct a link(2). So
     using this command in Org buffers will potentially create two
     links: a human-readable from the custom ID, and one that is
     globally unique and works even if the entry is moved from file to
     file.  Later, when inserting the link, you need to decide which one
     to use.
```

I guess it's a matter of preference.  For me, I just let generate the `ID`, and let Org give me the nice name in links and keep track of what file the actual heading is in.

**Comment by Fuco1** [comment]
Jumping by IDs is *massively* optimized in org mode.  It is instant, org itself keeps a cache of all the entries with ID property.  Going just by simple headline jumps is absolutely infeasible.

My org files together have over 30000 lines.  I am considering switching to this system but as I see it it will break quite soon because this would, among other things, create about 5000 files in a single directory which would totally kill performance.

**Comment by Fuco1** [comment]
The one really strong feature of this system is that it is all plaintext... using a database for something I might be using 20 years later makes me quite nervous... tech comes and goes but plaintext will be forever.  Even if org dies I can still at least grep it or build my own tools.

How about storing the parent/child relationship in the property drawers?  I think I would like that even more than the current `* Brainchildren` entry.  We can keep even bidirectional relations to speed things up even without caching.  Then write some graph query langauge on top and we're done.  The IDs for parents/children can be colon-delimited like `$PATH` which is easy to parse and easy to work with.  So all in all every headline would have three properties: `ID`, `PARENTS` and `CHILDREN`.

Now that I write it out, I have already built such a system, albeit for a completely different purpose (and it has no UI and is not very well optimized) but it works quite well: https://github.com/Fuco1/.emacs.d/blob/master/files/org-defs.el#L32 (It's for tracking dependencies between items)

(Well, bidi might blow the space usage quite a bit, but then we almost never form large cliques, so a case where we have 100 items each linked to each other as parent/child is pretty unrealistic... I assume most things will have 5-10 links at most which will keep things under a constant factor)

I would be in favour of respecting hierarchy in the files but we should be able to turn it off, either with a tag or a property.  It is quite simple to support multiple workflows like refiling or capture this way.  I use tons of capture templates or capture&refile under hierarchies where I get "free tags" so I get some organization immediately for free (a book is a book, I shouldn't need to link it every time I add a new book, just file it under the headline somewhere).

I'm not sure what to do about nested trees.  I feel it is right to add *all* parents as direct parents. A scifi book is a book so if I have a 2-level hierarchy I might want to jump right to a book. (because a "scifi" thing is not necessarily a book).  But since it is a scifi book it should also get that link.

```org
* Books
** Scifi
*** Foundation
  :PROPERTIES:
   :PARENTS: <link to asimov>
  :END:
** Fantasy
...
```

So in this setup `Foundation` should get three parents: `Books, `Scifi` and `Asimov`.

Well... this is quite fascinating to think about :)  I have spent *ages* and have used 4 systems now for organizing my org files (but every time it was an org scheme).  Years before org added taxonomies on tags I used that (I hacked it in) and had 1 enromous flat file (no nesting whatsoever) and stuff was handled with tag inheritance and ontology.  That didn't scale because Emacs can't really work a 50k line file :/

What I would love in the end is a "free-form" system (like `org-brain` now) which can cross-link anything ignoring the hierarchy but also supports headlines as @alphapapa suggests.  That way we can have relatively small amount of files so that it won't kill performance, we can have some locality for stuff like grep or tools that can query org files.

When I think about it this way it reminds me of wikipedia projects.  Things are organized under "Projects" (e.g. wikiproject mathematics https://en.wikipedia.org/wiki/Wikipedia:WikiProject_Mathematics) which gathers all things related (this would be one of our files) but you can of course link to anything from any maths article.

**Comment by Fuco1** [comment]
I went ahead and implemented a simple `org-graph` thing https://gist.github.com/Fuco1/98d2909097cdf65fb5ff6468a29906d5

Check the docstrings for how it works.

I will bolt some query language on top so you can say things like `(org-graph-get parents of X)` or `(org-graph-get siblings of X [by [parent Y])` and similar.

**Comment by Kungsgeten** [comment]
In the current version of `org-brain` there is some caching going on, implemented by @analyticd. Since then analyticd has switched to another system in his fork, using some kind of DB-file (or similar). His fork is focused on doing most of the work from `org-brain-visualize`, and I'm not sure if that is the way I want to go.

The main problem I see with storing parents in a file, or under a heading, is the risk of things going out of sync. Let's take the @Fuco1 example above, where `Foundation` have `Asimov` as a parent. When adding this parent, we'd also need to add `Foundation` as a child in the `Asimov` entry (which probably is in another file). And if we remove `Asimov` as a parent, then we have to remove `Foundation` as a child in the `Asimov` entry as well. If doing this with functions it is probably fine, but since it is all text there will be a problem if we remove the property by hand. We'd probably have to add a function to the save-hook and scan the file each time its saved in order to see if anything was removed.

Another option could ofcourse be not trying to sync stuff this way. `Foundation` could then have `Asimov` as a parent, but `Asimov` would not have `Foundation` as a child. That could probably be useful.

**Comment by Fuco1** [comment]
Well it goes automatically that the user should not edit the properties manually.  If we refile the entry to a different file org will probably lose track of it until the cache is rebuild (which takes not that long even on my database with 5000 IDs)

**Comment by Kungsgeten** [comment]
I have rewritten `org-brain` from scratch, please see: https://github.com/Kungsgeten/org-brain/tree/org-brain-id

This new version supports both files and ID headlines as entries. This will be merged into main in a couple of days, I'm just going to test it a bit more first.

**Comment by Fuco1** [comment]
I've been working on my own "implementation" mostly exploring various ways to encode graphs in org.  I think I will turn it into a generic "org graph" library which could in theory be used as a backend for many other similar projects.

I am still a bit "worried" about using files as entries.  Headlines are so much more rich in metadata and all the possibilities (you can also tag them, schedule, deadline, plan, TODO-state etc).  Maybe as a compromise I would make a file with a single headline, but I would definitely always group data under headlines.

**Comment by Kungsgeten** [comment]
It would definitely be easier not to treat files as entries. As it stands now, I've basically written two versions of most functions, because they have to treat files and headlines differently. I think allowing files in `org-brain` is a good thing though, since people like to structure their information in different ways. In a generic org graph library, that would probably not be needed (and it would work fine in `org-brain` too, to only have headlines as entries).

**Comment by Kungsgeten** [comment]
I have now merged the 0.4 branch into master, so it is now possible to work on both files and headlines as entries.
